docs/proposals-accepted/202304-query-path-tenancy.md

```diff
@@ -0,0 +1,95 @@
+
+2. Add the tenant identification as an optional field in the Store API protobuffer spec instead of an HTTP header.
+
+This isn't the preferred solution because the Store API isn't used all the way from the topmost components (e.g. Querier and Query Frontend) to the bottommost one (e.g. Store Gateway). The Query Frontend talks to the Querier via the Query API. A Querier can talk to other Queriers using the Query API via gRPC. Finally, a Querier talks to Store Gateways using the Store API. All these APIs would have to be modified and the transmission of this information would differ depending on the component and API being used. In addition to this complication, this implementation does not enable any capability that can't be achieved with HTTP header approach.
+
+### Alternative solutions
+
+1. Use a proxy to enforce and identify tenants in queries.
+
+While this could work for some of the features, like exporting per-tenant metrics, it would have to be inserted in front of many different components. Meanwhile, it doesn't solve the requirement for per-tenant configuration.
+
+2. Use a separate query path for each tenant.
+
+This incurs in a lot of wasted resources and demands manual work, unless a central tenant configuration is used and a controller is built around it to automatically manage the query paths.
+
+## Action Plan
+
+1. Query Frontend
+   1. Implement the `--query-frontend.tenant-header` (default is `""`") and `--query-frontend.default-tenant` (default is `"default-tenant"`) flags. Forward the tenant header to all the configured downstream query endpoints, if it is present.
+   2. **In the UI**, add a textbox where one can type a tenant name. If the textbox has any content, that will be sent using in the configured tenant header name.
+   3. Update the metrics exported by the Query Frontend to include a tenant label with the tenant indicated by the header.
+2. Querier
+   1. Implement the `--querier.tenant-header` (default is `""`") and `--querier.default-tenant` (default is `"default-tenant"`) flags. Forward the tenant header to all the downstream API calls (gRPC or HTTP, also using headers), if it is present.
+   2. Implement the `--querier.tenant-label-name` flag. If it is not empty, verify and enforce the tenant label in the query before it is handed to the query engine. As previously mentioned by the proposal, use prom-label-proxy's [Enforce.EnforceMatchers](https://github.com/prometheus-community/prom-label-proxy/blob/main/injectproxy/enforce.go#L141).
+   3. Update the metrics exported by the Querier to include a tenant label with the tenant indicated by the header.
+3. Store Gateway
+   1. Implement the `--store.tenant-header` (default is `""`") and `--query-frontend.default-tenant` (default is `"default-tenant"`) flags. Use the tenant header to identify the tenant in the metrics exported by the Store Gateway.
```

Comment:
Can't we use the tenant from request headers?

Reply:
How do we know the header in the request that includes the tenant's name if the header used to transmit this information can have its name changed? 

i.e. Query Frontend configured to use `X-User` as tenant header. How will Store Gateway know that `X-User` is the header being used for tenant identification and not anything else?

Reply:
The querier knows which header is used for tenancy, and it can propagate this header in either the request or in a special gRPC/http header.

Reply:
This overlaps with our conversation in the other comment, let's solve that one first and then we come back to this one if there will be anything left to discuss. 👍

Reply:
@bwplotka aahhh, that's good to know. I'll learn more about it and update the proposal accordingly. Thanks for the tip.